### DISCUSSION TITLE
Hasura Cron Trigger vs Linux Cron Jobs vs pg_cron vs Quartz

### DISCUSSION THREAD

**Discussion by ilbertt** [hasura/graphql-engine#8804] [General]
For the business logic I'm working on, I need to implement two types of batch jobs:
1. batch jobs that act directly on tables in database
2. batch jobs that fetch data from external REST endpoint and then act on tables in database

### Type 1.
I'm thinking of using [pg_cron](https://github.com/citusdata/pg_cron), also to avoid using any other database external service.

### Type 2.
Since I'm already exposing the PostgreSQL database with Hasura GraphQL engine, I'm evaluating Hasura Cron Triggers to implement those custom triggers. This could be fine for my structure, because I already have some basic dockerized [Gin Go endpoints](https://gin-gonic.com/) running in parallel with Hasura, hence it makes sense to also add other endpoints triggered by Hasura Cron Triggers.
However, I'm also evaluating writing cron jobs directly in Go with the help of [go-quartz](https://github.com/reugn/go-quartz) scheduling library or scheduling them in the container using [Linux cron](https://www.man7.org/linux/man-pages/man8/cron.8.html) directly.

What are your suggestion on the overall structure?
Thinking also from a CI/CD perspective (change scheduling on the fly, etc.):
- does it makes sense to have some batch jobs inside database and some other outside?
- does it makes sense to have them triggered by hasura or is it better to simply trigger them directly from their code?

**Comment by meetzaveri** [comment]
Hello ,

Thanks for your question.

re:
> does it makes sense to have some batch jobs inside database and some other outside?

It totally depends on your use case. If that batch jobs affects the database in any way, then you can have them outside.

re:
> does it makes sense to have them triggered by hasura or is it better to simply trigger them directly from their code?

Hasura provide cron triggers feature which can be managed from hasura console. The cron events data gets stored under `hdb_cron_events` table under `hdb_catalog` schema in your metadata database. The metadata database will be a seperate one from your main db in hasura engine.  Note - regardless of your data source (whether its mssql or postgres), the metadata database will be in postgres db (*which we spin up under the hood for hasura cloud and hasura CE).

Further, if you feel writing cron in go can be better according your use case, you can go ahead with that. Hasura cron trigger hits an endpoint every time its been triggered so you can also have your go endpoints (as you mentioned) and apply them as endpoint. It's upon you how flexible you want your structure.

Also if you can you elaborate on _**"is it better to simply trigger them directly from their code"**_ then it would be better to answer them too.

 Thanks again for dropping here in discussion.
